src/librustdoc/clean/types.rs

```diff
@@ -1396,15 +1396,7 @@ crate enum Type {
     BorrowedRef { lifetime: Option<Lifetime>, mutability: Mutability, type_: Box<Type> },
 
     /// A qualified path to an associated item: `<Type as Trait>::Name`
-    QPath {
-        name: Symbol,
-        self_type: Box<Type>,
-        /// FIXME: This is a hack that should be removed; see [this discussion][1].
-        ///
-        /// [1]: https://github.com/rust-lang/rust/pull/85479#discussion_r635729093
-        self_def_id: Option<DefId>,
-        trait_: Path,
-    },
+    QPath { name: Symbol, self_type: Box<Type>, should_show_cast: bool, trait_: Path },
```

Comment:
Can you make `should_show_cast` computed on-demand instead?

Reply:
I don't really see how to do it: we have three different cases and in two of them, we need to have access to an information we don't have anymore. But maybe you have an idea?

Reply:
I don't remember the context of the bug this fixed. Why shouldn't we show the "cast" for a qualified-path?

Reply:
We only show `Y as X` in case `X != Y` and that `Y` is a `self` type.

Reply:
Can't we recover that information from the `name`, `self_type`, and `trait_` fields then?

Reply:
The problem is that we need the `self_def_id` which is retrieved in four different ways:

```rust
let self_def_id = self_type.def_id(&cx.cache);
let self_def_id = cx.tcx.parent(self.def_id);
let self_def_id = DefId::local(qself.hir_id.owner.local_def_index);
let self_def_id = res.opt_def_id();
```

In two cases we can get it back (the two first), but for the others I don't really see how we can...

Reply:
Why is it computed in so many different ways? Can we consolidate some of them?

Reply:
It's because we group some slightly different things under `QPath`. So at the moment, I'm afraid we can't do much better... (Or at least I dont see how).